File: cpp/9202.cpp
#include <bits/stdc++.h>
#define fastio cin.tie(0)->sync_with_stdio(0)
using namespace std;

struct TrieNode {
	int nxt[26], idx;
	TrieNode() : nxt{}, idx() {}
};

struct Trie {
	vector<TrieNode> node;
	Trie() : node(1) {}

	void Add(const string& s, const int idx) {
		int cur = 0;
		for (const auto& c : s) {
			if (node[cur].nxt[c - 'A'] == 0) {
				node[cur].nxt[c - 'A'] = node.size();
				node.push_back(TrieNode());
			}
			cur = node[cur].nxt[c - 'A'];
		}
		node[cur].idx = idx;
	}
} T;

int n, q, visited[4][4];
string s[300'001], board[4];
vector<int> v;

void DFS(int x, int y, int pos) {
	if (T.node[pos].nxt[board[x][y] - 'A'] == 0) return;
	pos = T.node[pos].nxt[board[x][y] - 'A'], visited[x][y] = 1;
	if (T.node[pos].idx) v.push_back(T.node[pos].idx);
	for (int k = 0; k < 8; k++) {
		int nx = x + "10001222"[k] - '1';
		int ny = y + "22100012"[k] - '1';
		if (nx < 0 || nx >= 4 || ny < 0 || ny >= 4) continue;
		if (visited[nx][ny]) continue;
		DFS(nx, ny, pos);
	}
	visited[x][y] = 0;
}

int GetScore(int n) {
	return "00011235;"[n] - '0';
}

int main() {
	fastio;
	cin >> n;
	for (int i = 1; i <= n; i++) cin >> s[i], T.Add(s[i], i);
	cin >> q;
	while (q--) {
		v.clear();
		for (int i = 0; i < 4; i++) cin >> board[i];
		for (int i = 0; i < 4; i++) for (int j = 0; j < 4; j++) DFS(i, j, 0);
		sort(v.begin(), v.end()); v.erase(unique(v.begin(), v.end()), v.end());
		int score = 0, cnt = 0; string mx;
		for (const auto& i : v) {
			score += GetScore(s[i].size()), cnt++;
			if (mx.size() < s[i].size() || mx.size() == s[i].size() && mx > s[i]) mx = s[i];
		}
		cout << score << ' ' << mx << ' ' << cnt << '\n';
	}
}

File: python/9202.py
import sys
input = lambda: sys.stdin.readline().rstrip()

class TrieNode:
    def __init__(self):
        self.nxt = [0] * 26
        self.idx = -1

class Trie:
    def __init__(self):
        self.node = [TrieNode()]
    def Add(self, s, idx):
        cur = 0
        for c in s:
            if self.node[cur].nxt[ord(c) - 65] == 0:
                self.node[cur].nxt[ord(c) - 65] = len(self.node)
                self.node.append(TrieNode())
            cur = self.node[cur].nxt[ord(c) - 65]
        self.node[cur].idx = idx

n = int(input())
s = [input() for _ in range(n)]; input()

T = Trie()
board, v = [], {}
visited = [[0] * 4 for _ in range(4)]
for i in range(n): T.Add(s[i], i)

def DFS(x, y, pos):
    if T.node[pos].nxt[ord(board[x][y]) - 65] == 0: return
    pos = T.node[pos].nxt[ord(board[x][y]) - 65]; visited[x][y] = 1
    if T.node[pos].idx != -1: v.add(T.node[pos].idx)
    for dx, dy in ((0, 1), (-1, 1), (-1, 0), (-1, -1), (0, -1), (1, -1), (1, 0), (1, 1)):
        nx = x + dx; ny = y + dy
        if 0 <= nx < 4 and 0 <= ny < 4 and not visited[nx][ny]: DFS(nx, ny, pos)
    visited[x][y] = 0

def GetScore(n):
    if 3 <= n <= 4: return 1
    if n == 5: return 2
    if n == 6: return 3
    if n == 7: return 5
    if n == 8: return 11
    return 0

for _ in range(int(input())):
    board = [input() for _ in range(4)];
    v = set()
    for i in range(4):
        for j in range(4): DFS(i, j, 0)
    score, cnt, mx = 0, 0, ''
    for i in v:
        score += GetScore(len(s[i]))
        cnt += 1
        if len(mx) < len(s[i]) or len(mx) == len(s[i]) and mx > s[i]: mx = s[i]
    print(score, mx, cnt); input()
